source/agora/consensus/protocol/Nominator.d

```diff
@@ -377,56 +383,28 @@ extern(D):
 
     /***************************************************************************
 
-        Restore SCP's internal state based on the provided ledger state
-
-        Params:
-            ledger = the ledger instance
+        Restore SCP's internal state based on the stored latest envelopes
 
     ***************************************************************************/
 
-    private void restoreSCPState (Ledger ledger)
+    private void restoreSCPState ()
     {
-        import agora.common.Serializer;
-        import scpd.types.Stellar_SCP;
-        import scpd.types.Utils;
-        import scpd.types.Stellar_types : StellarHash = Hash, NodeID;
-        import std.range;
-
-        auto pub_key = NodeID(uint256(this.key_pair.address));
-
-        auto block = ledger.getBlocksFrom(ledger.getBlockHeight()).front;
-        auto serialized = block.serializeFull();
-        auto vec = serialized.toVec();
-        auto value = duplicate_value(&vec);
-        SCPStatement statement =
+        foreach (const ref SCPEnvelope env; this.latest_envelopes)
         {
-            nodeID: pub_key,
-            slotIndex: block.header.height,
-            pledges: {
-                type_: SCPStatementType.SCP_ST_EXTERNALIZE,
-                externalize_: {
-                    commit: {
-                        counter: 0,
-                        value: value,
-                    },
-                    nH: 0,
-                },
-            },
-        };
-
-        SCPEnvelope env = SCPEnvelope(statement);
-        this.scp.setStateFromEnvelope(block.header.height, env);
-        if (!this.scp.isSlotFullyValidated(block.header.height))
-            assert(0);
+            this.scp.setStateFromEnvelope(env.statement.slotIndex, env);
+            if (!this.scp.isSlotFullyValidated(env.statement.slotIndex))
+                assert(0);
+        }
 
         // there should at least be a genesis block
-        if (this.scp.empty())
-            assert(0);
+        // if (this.scp.empty())
+        //     assert(0);
```

Comment:
This verification is essential?

Reply:
This can be removed. For the genesis block we won't have any state in SCP yet, and therefore this check is unnecessary.